frappe/integrations/doctype/paytm_settings/paytm_settings.py

```diff
@@ -0,0 +1,166 @@
+		if currency not in self.supported_currencies:
+			frappe.throw(_("Please select another payment method. Stripe does not support transactions in currency '{0}'").format(currency))
+
+	def get_payment_url(self, **kwargs):
+		'''Return payment url with several params'''
+		# create unique order id by making it equal to the integration request
+		integration_request = create_request_log(kwargs, "Host", "Paytm")
+		kwargs.update(dict(order_id=integration_request.name))
+
+		return get_url("./integrations/paytm_checkout?{0}".format(urlencode(kwargs)))
+
+def get_paytm_config():
+	''' Returns paytm config '''
+
+	paytm_config = frappe.db.get_singles_dict('Paytm Settings')
+	paytm_config.update(dict(merchant_key=get_decrypted_password('Paytm Settings', 'Paytm Settings', 'merchant_key')))
+
+	if cint(paytm_config.staging):
+		paytm_config.update(dict(
+			website="WEBSTAGING",
+			url='https://securegw-stage.paytm.in/order/process',
+			transaction_status_url='https://securegw-stage.paytm.in/order/status',
+			industry_type_id='RETAIL'
+		))
+	else:
+		paytm_config.update(dict(
+			url='https://securegw.paytm.in/order/process',
+			transaction_status_url='https://securegw.paytm.in/order/status',
+		))
+	return paytm_config
+
+def get_paytm_params(payment_details, order_id, paytm_config):
+
+	# initialize a dictionary
+	paytm_params = dict()
+	
+	redirect_uri = get_request_site_address(True) + "/api/method/frappe.integrations.doctype.paytm_settings.paytm_settings.verify_transaction"
+
+
+	paytm_params.update({
+		"MID" : paytm_config.merchant_id,
+		"WEBSITE" : paytm_config.website,
+		"INDUSTRY_TYPE_ID" : paytm_config.industry_type_id,
+		"CHANNEL_ID" : "WEB",
+		"ORDER_ID" : order_id,
+		"CUST_ID" : payment_details['payer_email'],
+		"EMAIL" : payment_details['payer_email'],
+		"TXN_AMOUNT" : cstr(flt(payment_details['amount'], 2)),
+		"CALLBACK_URL" : redirect_uri,
+	})
+
+	checksum = generateSignature(paytm_params, paytm_config.merchant_key)
+
+	paytm_params.update({
+		"CHECKSUMHASH" : checksum
+	})
+
+	return paytm_params
+
+@frappe.whitelist(allow_guest=True)
+def verify_transaction(**kwargs):
+	'''Verify checksum for received data in the callback and then verify the transaction'''
+	paytm_config = get_paytm_config()
+	received_data = frappe._dict(kwargs)
+	is_valid_checksum = False
+
+	paytm_params = {}
+	for key, value in received_data.items(): 
+		if key == 'CHECKSUMHASH':
+			paytm_checksum = value
+		elif key == 'cmd':
+			continue
+		else:
+			paytm_params[key] = value
+
+	if paytm_params and paytm_config and paytm_checksum:
+		# Verify checksum
+		is_valid_checksum = verifySignature(paytm_params, paytm_config.merchant_key, paytm_checksum)
+
+	if is_valid_checksum and received_data['RESPCODE'] == '01':
+		verify_transaction_status(paytm_config, received_data['ORDERID'])
+	else:
+		frappe.respond_as_web_page("Payment Failed",
+			"Transaction failed to complete. In case of any deductions, deducted amount will get refunded to your account.",
```

Comment:
maybe a link to the paytm refund policies in here would be good?

Reply:
link keeps on changing, merchant can handle this issues directly.